tower/src/limit/concurrency/layer.rs

```diff
@@ -1,24 +1,67 @@
+use std::sync::Arc;
+
 use super::ConcurrencyLimit;
+use tokio::sync::Semaphore;
 use tower_layer::Layer;
 
 /// Enforces a limit on the concurrent number of requests the underlying
 /// service can handle.
 #[derive(Debug, Clone)]
-pub struct ConcurrencyLimitLayer {
-    max: usize,
+pub struct ConcurrencyLimitLayer<T = usize> {
+    semaphore: T,
 }
 
 impl ConcurrencyLimitLayer {
     /// Create a new concurrency limit layer.
     pub fn new(max: usize) -> Self {
-        ConcurrencyLimitLayer { max }
+        ConcurrencyLimitLayer { semaphore: max }
+    }
+}
+
+impl<T> ConcurrencyLimitLayer<T>
+where
+    T: ToOwnedSemaphore,
+{
+    /// Create a new concurrency limit layer from a type that implements [`ToOwnedSemaphore`]
+    pub fn new_with_semaphore(semaphore: T) -> Self {
+        ConcurrencyLimitLayer { semaphore }
     }
 }
 
-impl<S> Layer<S> for ConcurrencyLimitLayer {
+impl<S, T> Layer<S> for ConcurrencyLimitLayer<T>
+where
+    T: ToOwnedSemaphore,
+{
     type Service = ConcurrencyLimit<S>;
 
     fn layer(&self, service: S) -> Self::Service {
-        ConcurrencyLimit::new(service, self.max)
+        ConcurrencyLimit::new(service, &self.semaphore)
+    }
+}
+
+/// Produces an owned [`Semaphore`] (`Arc<Semaphore>`)
+pub trait ToOwnedSemaphore {
+    /// Returns an owned [`Semaphore`] for the type
+    fn to_owned_semaphore(&self) -> Arc<Semaphore>;
```

Comment:
AFAICT, this trait will basically only ever be implemented for `usize` and `Arc<Semaphore>`, right? It doesn't seem like there's a lot that users could do with implementations for other types? 

Hypothetically, I could maybe see a situation where the concurrency limiting policy depends on the service being wrapped in some way, as a use-case for custom implementations, but the current trait doesn't make that possible.  Perhaps it could take a reference to the inner service when producing the semaphore...but I'm still not sure if something like that would actually be useful.

I would generally try to avoid exposing a trait in the public API if users aren't going to be implementing themselves it or using it as a trait bound, if that's possible. It feels like the user-facing complexity could be avoided if we just had two separate layers.